### PR TITLE
Redefining numbers in Quack

### PR DESCRIPTION
**Pull request by luizperes**
This PR makes Quack numbers consistent by changing some old rules. In this way, we do not need to use PHP notation anymore.

There are numerous inconsistencies in PHP such as:
  - Hexadecimal notation can be represented by `0xFFfff0` and `0XFFfff0`
  - Binary notation is only represented by `0b0001` (`0B0001` is not valid).
  - Octal notation is weird: `0666` (__?__) `// isn't that a decimal notation number?`

This PR is redefining Quack numbers to:

|              | Decimal |  Binary | Octal | Hexadecimal  | Double | Double Exp |
| :--------: | :---------: | :----------: | :------: | :------: | :----------------: | :--------------: |
| __Then__  |  2222   | 0`[bB]`10011010  | 0765  | 0`[xX]`FA19  |  3.1415  |  - |
| __Now__  |   2222 |  0`b`10011010 |  0`o`765   | 0`x`FA19  |  3.1415  | 2.15`e[+-]`10 |

In order to ease our future tasks of analysing number types (such as comparison optimizations and reducing the complexity of always checking a number), I included new `Tag` types: `T_INT_OCT`, `T_INT_BIN`, `T_INT_HEX` and `T_DOUBLE_EXP`. Thus, I also created a member named `notation` inside the `NumberExpr` class. Having said that, `octal`, `binary` and `hexadecimal` notation numbers are still __int__ and `scientific` notation ones are __double__.

Obs.: If we took `Quack` only as a Transpiler, we wouldn't need to check those numbers, however, the Quack project is big and I am thinking about the future implications of not having those tags. 

------------------------------------------------------------------
- `digit()` is now using the `KMP` algorithm approach for `octal`, `binary` and `hexadecimal` notations.

 - The second part (checking for floating numbers) uses a basic state machine:

photo435744009096374236

### CODE REVIEW

File: src/lexer/Tokenizer.php [hunk 1]
```diff
@@ -82,79 +82,89 @@ public function nextToken()
...
-            while ($index < $current_buffer_size && in_array($buffer[$index], range(0, 7))) {
-                $oct_buffer[] = $buffer[$index];
-                $index++;
-            }
-
-            $value = implode($oct_buffer);
-            $this->column += $current_buffer_size;
-            return new Token(Tag::T_INTEGER, $this->symbol_table->add($value));
+        // check optional exp state: looking for a 'e+' or 'e-' and integers
+        if (!$this->isEnd() && ($this->matches('e+') || $this->matches('e-'))
```
Comment (haskellcamargo) [inline src/lexer/Tokenizer.php]: When only `e` is found, we also have a valid positive form.

**Review by haskellcamargo — CHANGES_REQUESTED** [review]

File: src/lexer/Tokenizer.php (hunk 1 above)
Comment (luizperes) [reply src/lexer/Tokenizer.php]: will do it!

Comment (luizperes) [reply src/lexer/Tokenizer.php]: I have just done it! Please check that out @haskellcamargo

**Review by haskellcamargo — APPROVED** [review]

**Comment by haskellcamargo** [comment]
Definitely LGTM!

**Comment by luizperes** [comment]
yay!

**Pull request merged**